src/librustc/ty/print/pretty.rs

```diff
@@ -1530,7 +1530,7 @@ define_print_and_forward_display! {
     &'tcx ty::Const<'tcx> {
         match self.val {
             ConstValue::Unevaluated(..) |
-            ConstValue::Infer(..) => p!(write("_")),
+            ConstValue::Infer(..) => p!(write("_: {}", self.ty)),
```

Comment:
This sort of thing is very bad for `ty::pretty`, sadly there's no way I can think of to turn it into a compiler error...

A correct version would be `p!(write("_: "), print(self.ty))`.

Oh, also, `p!(write("{:?}", self))` below will also end up printing typesystem entities via `fmt::Debug` (which may differ from the current pretty-printer)

Was that meant to be `write("{:?}", self.val)` instead?

Reply:
cc @varkor

Reply:
we already have a best effort pretty printer for constants in https://github.com/rust-lang/rust/blob/070cebd0aa76702eaaf7c7f340615805f661e937/src/librustc/mir/mod.rs#L2703 which is what debug printing `&'tcx ty::Const<'tcx>` triggers. We should totally unify this

Reply:
@oli-obk Oh, all that stuff definitely needs to implement the `ty::print::Print` API, for it to be *usable at all* from pretty-printing.

Reply:
It does now